src/solidity/decodeInfo.js

```diff
@@ -0,0 +1,353 @@
+  * @param {Object} stateDefinitions  - all state definition given by the AST (including struct and enum type declaration)
+  * @return {Array} containing all members of the current struct type
+  */
+function getStructMembers (typeName, stateDefinitions) {
+  var members = []
+  var storageBytes = 0
+  for (var k in stateDefinitions) {
+    var dec = stateDefinitions[k]
+    if (dec.name === 'StructDefinition' && typeName.indexOf('struct ' + dec.attributes.name) === 0) {
+      for (var i in dec.children) {
+        var member = dec.children[i]
+        var decoded = decode(member.attributes.type, stateDefinitions)
+        members.push(decoded)
+        if (decoded.needsFreeStorageSlot) {
+          storageBytes = Math.ceil(storageBytes / 32) * 32
+        }
+        storageBytes += decoded.storageBytes
+      }
+      break
+    }
+  }
+  return {
+    members: members,
+    storageBytes: storageBytes
+  }
+}
+
+/**
+  * return the size of the current array
+  *
+  * @param {String} typeName - short type ( e.g uint[][4] )
+  * @return {String|Int} return 'dynamic' if dynamic array | return size of the array
+  */
+function extractArraySize (typeName) {
+  if (typeName.indexOf('[') !== -1) {
+    var squareBracket = /\[([0-9]+|\s*)\]/g
+    var dim = typeName.match(squareBracket)
+    var size = dim[dim.length - 1]
+    if (size === '[]') {
+      return 'dynamic'
+    } else {
+      return parseInt(dim[dim.length - 1].replace('[', '').replace(']'))
+    }
+  }
+}
+
+/**
+  * extract the underlying type
+  *
+  * @param {String} fullType - type given by the AST (ex: uint[2] storage ref[2])
+  * @return {String} return the first part of the full type. don not keep the array declaration ( uint[2] storage ref[2] will return uint)
+  */
+function extractUnderlyingType (fullType) {
+  var splitted = fullType.split(' ')
+  if (fullType.indexOf('enum') === 0 || fullType.indexOf('struct') === 0) {
+    return splitted[0] + ' ' + splitted[1]
+  }
+  if (splitted.length > 0) {
+    fullType = splitted[0]
+  }
+  if (fullType[fullType.length - 1] === ']') {
+    return fullType.substring(0, fullType.indexOf('['))
+  }
+  return fullType
+}
+
+/**
+  * get the array dimensions
+  *
+  * @param {String} fullType - type given by the AST
+  * @return {Array} containing all the dimensions and size of the array (e.g ['[3]', '[]'] )
+  */
+function extractDimensions (fullType) {
+  var ret = []
+  if (fullType.indexOf('[') !== -1) {
+    var squareBracket = /\[([0-9]+|\s*)\]/g
+    var dim = fullType.match(squareBracket)
+    return dim
+  }
+  return ret
+}
+
+/**
+  * parse the full type
+  *
+  * @param {String} fullType - type given by the AST (ex: uint[2] storage ref[2])
+  * @return {String} returns the token type (used to instanciate the right decoder) (uint[2] storage ref[2] will return 'array', uint256 will return uintX)
+  */
+function extractTokenType (fullType) {
+  if (fullType.indexOf('[') !== -1) {
```

Comment:
also yes but afak if the type given by the ast contains `[` that's an array no? or I miss something